eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Recipient.scala

```diff
@@ -0,0 +1,162 @@
+
+  /** Edges that aren't part of the public graph and can be used to reach the recipient. */
+  def extraEdges: Seq[Invoice.ExtraEdge]
+
+  /** Build a payment to the recipient using the route provided. */
+  def buildPayloads(paymentHash: ByteVector32, route: Route): Try[PaymentPayloads]
+}
+
+/** A payment recipient that can directly be found in the routing graph. */
+case class ClearRecipient(nodeId: PublicKey,
+                          features: Features[InvoiceFeature],
+                          totalAmount: MilliSatoshi,
+                          expiry: CltvExpiry,
+                          paymentSecret: ByteVector32,
+                          extraEdges: Seq[Invoice.BasicEdge] = Nil,
+                          paymentMetadata_opt: Option[ByteVector] = None,
+                          nextTrampolineOnion_opt: Option[OnionRoutingPacket] = None,
+                          customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
+  override def buildPayloads(paymentHash: ByteVector32, route: Route): Try[PaymentPayloads] = {
+    val finalPayload = nextTrampolineOnion_opt match {
+      case Some(trampolinePacket) => NodePayload(nodeId, FinalPayload.Standard.createTrampolinePayload(route.amount, totalAmount, expiry, paymentSecret, trampolinePacket))
+      case None => NodePayload(nodeId, FinalPayload.Standard.createPayload(route.amount, totalAmount, expiry, paymentSecret, paymentMetadata_opt, customTlvs))
+    }
+    Success(OutgoingPaymentPacket.buildPayloads(route.amount, expiry, finalPayload, route.hops))
+  }
+}
+
+object ClearRecipient {
+  def apply(invoice: Bolt11Invoice, totalAmount: MilliSatoshi, expiry: CltvExpiry, customTlvs: Seq[GenericTlv]): ClearRecipient = {
+    ClearRecipient(invoice.nodeId, invoice.features, totalAmount, expiry, invoice.paymentSecret, invoice.extraEdges, invoice.paymentMetadata, None, customTlvs)
+  }
+}
+
+/** A payment recipient that doesn't expect to receive a payment and can directly be found in the routing graph. */
+case class SpontaneousRecipient(nodeId: PublicKey,
+                                amount: MilliSatoshi,
+                                expiry: CltvExpiry,
+                                preimage: ByteVector32,
+                                customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
+  override val totalAmount = amount
+  override val features = Features.empty
+  override val extraEdges = Nil
+
+  override def buildPayloads(paymentHash: ByteVector32, route: Route): Try[PaymentPayloads] = {
+    val finalPayload = NodePayload(nodeId, FinalPayload.Standard.createKeySendPayload(route.amount, amount, expiry, preimage, customTlvs))
+    Success(OutgoingPaymentPacket.buildPayloads(amount, expiry, finalPayload, route.hops))
+  }
+}
+
+sealed trait TrampolineRecipient extends Recipient {
+  // @formatter:off
+  def trampolineNodeId: PublicKey
+  def trampolineFees: MilliSatoshi
+  def trampolineAmount: MilliSatoshi
+  def trampolineExpiry: CltvExpiry
```

Comment:
`trampolineNodeId`, `trampolineAmount` and `trampolineExpiry` should just be `nodeId`, `totalAmount` and `expiry`.

Reply:
I disagree, we would be losing information about the recipient in that case. The recipient of the payment is *not* the first trampoline node, it's the last node in the route / issuer of the invoice (same for amounts and invoice). We can have the best of both worlds in a future PR that refactors trampoline, as mentioned above, but for now I think it's better to temporarily keep this (which is actually similar to the previous code where this was done inside the `PaymentInitiator`) and improve it step-by-step.

Reply:
We don't need that information, it's not used anywhere outside of `Recipient`.
Maybe the problem here is that you're not sure about the role of the `Recipient` trait. I think it should only contain the node id to send the payment to (which in the case of the trampoline payment is the trampoline id, not the final recipient id), how to reach it if it's a private node (extra edges) and a function to build the payloads, that's it. `features`, `expiry` and even `totalAmount` are not needed there, they should obviously be part of the class so that `buildPayloads` can use them but they shouldn't be in the trait. And actually that's already how you use `Recipient`, the fields `features` and `expiry` are never accessed and `totalAmount` is only accessed in `MultiPartPaymentLifecycle` because it was historically read from the final payload because it happens to be there, but it could be put inside `SendMultiPartPayment` as is done in `PaymentLifecycle` (`amount` is in `SendPayment`).
Here is a commit that does it (except removing `totalAmount`): https://github.com/ACINQ/eclair/commit/9730ef08e662497d2e3c671b18d65abab420d70d

Reply:
> We don't need that information, it's not used anywhere outside of Recipient.

Of course we need it, it's important to remember who the recipient of a payment is for accounting. The reason you think it's not used anywhere is because it's duplicated inside the `SendPaymentConfig` object, which is an ugly hack that we'll now be able to clean up: when refactoring trampoline payments, we'll be able to remove `recipientAmount` and `recipientNodeId` from `SendPaymentConfig`, since they naturally belong to the `Recipient` abstraction.

> Maybe the problem here is that you're not sure about the role of the Recipient trait. I think it should only contain the node id to send the payment to

Again, doing that would force us to lose information about the recipient (or force us to keep the `SendPaymentConfig` hack). A payment recipient is quite a natural abstraction, and it should really contain values for what the recipient is (its `nodeId`, how to reach it) and what it will receive (total amount, expiry, etc).

You're right that `features` is currently unused and can be removed, I'll do that. I introduced it initially for the trampoline-to-legacy case, but I ended up keeping the whole invoice. We can always re-add it later if necessary.

I think you're getting confused because this PR doesn't include the trampoline refactoring, so it's hard to see what comes next. I'll try to get that refactoring done and built on top of this PR soon, it will help verify that the abstraction works correctly and why it's useful.

Reply:
I added the `targetNodeId` to the `Recipient` abstraction, which is a bit cumbersome but I don't see how else we can do it since trampoline MPP prevents us from only relying on the path-finding algorithm without modifications.